django/db/models/sql/query.py

```diff
@@ -569,6 +569,19 @@ def combine(self, rhs, connector):
         if self.distinct_fields != rhs.distinct_fields:
             raise TypeError('Cannot combine queries with different distinct fields.')
 
+        # If `lhs` and `rhs` shares the same alias prefix, it is possible to have
+        # conflicting alias changes like T4 -> T5 and T5 -> T6 (which ends up as
+        # T4 -> T6) while combining two querysets. To prevent this, change alias
+        # prefix of the `rhs`.
+        if self.alias_prefix == rhs.alias_prefix:
+            old_alias_prefix = rhs.alias_prefix
+            rhs.bump_prefix(self, change_aliases=False)
+            rhs.change_aliases({
+                alias: alias.replace(old_alias_prefix, rhs.alias_prefix)
+                for alias, join in rhs.alias_map.items()
+                if join.table_alias != join.table_name
```

Comment:
This part might be refactored, but I wasn't sure if using `rhs.table_map` would be better or not.

Reply:
In the end the algorithm below that generates `change_map` has a single requirement with regards to alias; the initial alias must be shared.

I think it would be great if we could reuse `bump_prefix`'s usage of `change_aliases` while making sure that this particular initial alias (`self.get_initial_alias()`) is not _bumped_.